fhir-persistence-schema/docs/DB2MultiTenancy.md

```diff
@@ -0,0 +1,140 @@
+
+### Schema: Administrative
+
+The administrative schema's name is FHIR_ADMIN.  FHIR_ADMIN has 
+- three tables: TENANTS, TENANT_KEYS, VERSION_HISTORY
+- one sequence: FHIR_ADMIN.TENANT_SEQUENCE
+- one stored procedure: SET_TENANT
+
+The tables have supporting indicies and privileges. 
+
+**Table: TENANTS**
+
+The Tenants table maps a given tenant to the Multi-Tenant ID (MT_ID) for the given tenant.  The MT_ID is subsequently used to restrict row access based on the matching value. 
+
+| Column Name | Data Type | Length | Null | Purpose |
+|----------|---------------|---------------|---------------|---------------------------------------------|
+| MT_ID | Integer | 4 | No | Unique ID identifying the Tenant's ID|
+| TENANT_NAME | VARCHAR | 36 | No | The name of the tenant |
+| TENANT_STATUS | VARCHAR | 16 | No | The status of the Tenant|
+
+MT_ID is the value used to assign ROW Access Control in the Multi-Tenant schema. The MT_ID is assigned from the FHIR_ADMIN.TENANT_SEQUENCE.  The value is to 1 to 9999, with no cycles. 
+
+Tenant Status is one of the following: 
+| Status | Purpose |
+|----------|---------------------------------------------|
+| PROVISIONING | Indicates the tenant schema is in the process of adding a set of tenant tables/partitions |
+| ALLOCATED | Indicates the process of creating the tenant schema is finished, and the schema is ready for use. Only allocated tenants are available for use in the `fhir-persistence-jdbc` code |
+| FROZEN | Indicates the previous tenant data is not to be used, and the tenant data is in the process of being removed |
+| DROPPED | Indicates the previous tenant data is now removed |
+| FREE | Indicates a set of tables is provisioned, and available to be allocated to a specific tenant |
+
+**Table: TENANT_KEYS**
+
+The Tenant Keys table stores a hashed version of the tenant specific key. Upon creating a session and calling the SET_TENANT procedure, the SET_TENANT queries the incoming key against the TENANT_KEYS database to map to a specific tenant MT_ID. 
+
+| Column Name | Data Type | Length | Null | Purpose |
+|----------|---------------|---------------|---------------|---------------------------------------------|
+| TENANT_KEY_ID | Integer | 4 | No | The ID is from the sequence, and is used to identify the unique entry in the row |
+| MT_ID | Integer | 4 | No | Matches to the MT_ID in the TENANTS table |
+| TENANT_SALT | VARCHAR | 44 | No | A 32 byte random value used as a salt when computing the SHA-256 tenant hash |
+| TENANT_HASH | VARBINARY | 32 | No | Each tenant's data is encrypted. |
+
+The tenant_key is a 32 byte random value created when the tenant is provisioned. This is a secret, and must be passed by the client to the database in order to obtain access to a tenant. A tenant may have multiple tenant keys to support key rotation. An example of the tenant key is `GQUHRi0ETFw0Xe+p1VtVDki4JFZs885bGM35hXRyq/E=`
+
```

Comment:
how to rotate the keys?  does this mean we can run the com.ibm.fhir.schema.app.Main app multiple times to add new tenant for the same tenant name? and these multiple keys can work at the same time...

Reply:
That's right. You create a new one, and you can remove the unused one.